File: cpp/grade-school/grade_school.cpp
#include "grade_school.h"

std::map<int, std::vector<std::string>> grade_school::school::roster() {
  std::map<int, std::vector<std::string>> rosterCopy = m_roster;
  return rosterCopy;
}

void grade_school::school::add(std::string name, int grade) {
  m_roster[grade].push_back(name);
  std::sort(m_roster[grade].begin(), m_roster[grade].end());
}

std::vector<std::string> grade_school::school::grade(int p_grade) {
  if (!roster().empty()) {
    std::vector<std::string> grades = m_roster[p_grade];
    std::sort(grades.begin(), grades.end());
    return grades;
  } else {
    std::vector<std::string> emptyVector;
    return emptyVector;
  }
}

File: javascript/grade-school/grade-school.js
export class GradeSchool {
  constructor(roster = {}) {
    this._roster = roster;
  }

  roster() {
    return JSON.parse(JSON.stringify(this._roster));
  }

  add(name, grade) {
    if (this._roster[grade]) {
      this._roster[grade].push(name);
      this._roster[grade].sort();
    } else {
      this._roster[grade] = [ name ];
    }
  }

  grade(grade) {
    return this._roster[grade]
      ? [ ...this._roster[grade] ]
      : [];
  }
}
